internal/pkg/cli/task_run.go

```diff
@@ -134,6 +175,172 @@ func (o *runTaskOpts) Ask() error {
+	if o.image == "" && o.dockerfilePath == "" {
+		o.dockerfilePath = defaultDockerfilePath
+	}
+
+	if o.imageTag == "" {
+		o.imageTag = defaultImageTag
+	}
+
+	if err := o.getNetworkConfig(); err != nil {
+		return err
+	}
+
+	cluster, err := o.getCluster()
+	if err != nil {
+		return err
+	}
+
+	if err := o.deployTaskResources(); err != nil {
+		return err
+	}
+
+	//if image is not provided, then we build the image and push to ECR repo
+	if o.image == "" {
+		uri, err := o.pushToECRRepo()
+		if err != nil {
+			return err
+		}
+		o.image = fmt.Sprintf(fmtImageURL, uri, o.imageTag)
+
+		// update image to stack
+		if err := o.updateTaskResources(); err != nil {
+			return err
+		}
+	}
+
+	return o.runTask(cluster)
+}
+
+func (o *runTaskOpts) getCluster() (string, error) {
+	if o.env == config.EnvNameNone {
+		cluster, err := o.ecsGetter.DefaultCluster()
+		if err != nil {
+			return "", fmt.Errorf("get default cluster: %w", err)
+		}
+		return cluster, nil
+	}
+
+	clusters, err := o.resourceGetter.GetResourcesByTags(clusterResourceType, map[string]string{
+		deploy.AppTagKey: o.AppName(),
+		deploy.EnvTagKey: o.env,
+	})
+
+	if err != nil {
+		return "", fmt.Errorf("get cluster by env %s: %w", o.env, err)
+	}
+
+	if len(clusters) == 0 {
+		return "", fmt.Errorf("no cluster found under env %s in app %s", o.env, o.AppName())
+	}
+	return clusters[0], nil
+}
+
+func (o *runTaskOpts) runTask(cluster string) error {
+	o.spinner.Start(fmt.Sprintf("Waiting for %s to be running.", o.groupName))
+	err := o.starter.RunTask(ecs.RunTaskInput{
+		Cluster:        cluster,
+		Count:          o.count,
+		Subnets:        o.subnets,
+		SecurityGroups: o.securityGroups,
+		TaskFamilyName: fmt.Sprintf(fmtTaskFamilyName, o.groupName),
+	})
+
+	if err != nil {
+		o.spinner.Stop(log.Serrorln(fmt.Sprintf("Failed to run %s", o.groupName)))
+		return fmt.Errorf("run task %s: %w", o.groupName, err)
+	}
+
+	o.spinner.Stop(log.Ssuccessln(fmt.Sprintf("Task(s) %s is running.", o.groupName)))
+	return nil
+}
+
+func (o *runTaskOpts) getNetworkConfig() error {
+	if o.subnets == nil {
+		subnets, err := o.vpcGetter.GetSubnetIDs(o.AppName(), o.env)
+		if err != nil {
+			return fmt.Errorf("get subnet IDs from environment %s: %w", o.env, err)
+		}
+		o.subnets = subnets
+	}
+
+	// if the task is to be run in None environment (default environment), leave security groups as nil since it is not
+	// a required field for run task
+	if o.securityGroups == nil {
+		securityGroups, err := o.vpcGetter.GetSecurityGroups(o.AppName(), o.env)
+		if err != nil {
+			return fmt.Errorf("get security groups from environment %s: %w", o.env, err)
+		}
+		o.securityGroups = securityGroups
+	}
+	return nil
+}
+
+func (o *runTaskOpts) deployTaskResources() error {
+	o.spinner.Start(fmt.Sprintf("Deploying resources for task %s", color.HighlightUserInput(o.groupName)))
+	if err := o.deploy(); err != nil {
+		var errChangeSetEmpty *awscfn.ErrChangeSetEmpty
+		if !errors.As(err, &errChangeSetEmpty) {
```

Comment:
I wonder if this logic should live inside of the `deploy.DeployTask` method, what do you think?

My dream state 💭 for any `cli` command is for it just delegate work to other packages. It should ideally just contain a bunch of method calls in `Execute` and no custom logic.

Reply:
It can definitely be moved to `deploy.DeployTask`. When I put that logic to `Execute`, my concern was that perhaps it's not necessarily true that all calls to `DeployTask` want to handle `errChangeSetEmpty` the same way. But yes I agree with the dream state 💭 If any future call to `DeployTask` want to handle it differently, there are other ways to deal with that.